--- a/drivers/gpu/drm/drm_fb_helper.c	2021-08-26 14:51:21.000000000 +0200
+++ b/drivers/gpu/drm/drm_fb_helper.c	2022-07-22 08:10:47.713629624 +0200
@@ -438,8 +438,8 @@
 	 *       drm_format_info. Then derive all other values from the
 	 *       format.
 	 */
-	if (!preferred_bpp)
-		preferred_bpp = 32;
+	 
+	preferred_bpp = 32;
 
 	INIT_LIST_HEAD(&helper->kernel_fb_list);
 	spin_lock_init(&helper->damage_lock);
