--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
@@ -604,7 +604,7 @@
 		if (PTR_ERR(plat->stmmac_rst) == -EPROBE_DEFER)
 			goto error_hw_init;
 
-		dev_info(&pdev->dev, "no reset control found\n");
+		//dev_info(&pdev->dev, "no reset control found\n");
 		plat->stmmac_rst = NULL;
 	}
 
@@ -681,7 +681,7 @@
 	if (stmmac_res->lpi_irq < 0) {
 		if (stmmac_res->lpi_irq == -EPROBE_DEFER)
 			return -EPROBE_DEFER;
-		dev_info(&pdev->dev, "IRQ eth_lpi not found\n");
+		//dev_info(&pdev->dev, "IRQ eth_lpi not found\n");
 	}
 
 	stmmac_res->addr = devm_platform_ioremap_resource(pdev, 0);
