--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c	2026-04-16 02:12:51.764534809 +0200
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c	2026-04-18 00:15:08.257911087 +0200
@@ -773,7 +773,6 @@
 	if (stmmac_res->lpi_irq < 0) {
 		if (stmmac_res->lpi_irq == -EPROBE_DEFER)
 			return -EPROBE_DEFER;
-		dev_info(&pdev->dev, "IRQ eth_lpi not found\n");
 	}
 
 	stmmac_res->addr = devm_platform_ioremap_resource(pdev, 0);
