--- a/env/common.c
+++ b/env/common.c
@@ -122,8 +122,13 @@
 		memcpy(&crc, &ep->crc, sizeof(crc));
 
 		if (crc32(0, ep->data, ENV_SIZE) != crc) {
-			env_set_default("bad CRC", 0);
-			return -ENOMSG; /* needed for env_load() */
+			env_set_default(NULL, 0);
+			
+			if (env_set("boot_device", "nand")) {
+				debug("Failed to set boot_device\n");
+			}
+
+			return 0;
 		}
 	}
 
@@ -168,7 +173,13 @@
 				tmp_env2->crc;
 
 	if (!crc1_ok && !crc2_ok) {
-		return -ENOMSG; /* needed for env_load() */
+		env_set_default(NULL, 0);
+			
+		if (env_set("boot_device", "nand")) {
+			debug("Failed to set boot_device\n");
+		}
+
+		return 0;
 	} else if (crc1_ok && !crc2_ok) {
 		gd->env_valid = ENV_VALID;
 	} else if (!crc1_ok && crc2_ok) {
@@ -253,7 +264,12 @@
 		env_set_default(NULL, 0);
 #else
 		bootstage_error(BOOTSTAGE_ID_NET_CHECKSUM);
-		env_set_default("bad CRC", 0);
+				
+		env_set_default(NULL, 0);
+			
+		if (env_set("boot_device", "nand")) {
+			debug("Failed to set boot_device\n");
+		}
 #endif
 	} else {
 		env_load();
