--- a/env/common.c
+++ b/env/common.c
@@ -122,7 +122,7 @@
 		memcpy(&crc, &ep->crc, sizeof(crc));
 
 		if (crc32(0, ep->data, ENV_SIZE) != crc) {
-			env_set_default("bad CRC", 0);
+			env_set_default(NULL, 0);
 			return -ENOMSG; /* needed for env_load() */
 		}
 	}
@@ -203,7 +203,7 @@
 		env_set_default("bad env area", 0);
 		return -EIO;
 	} else if (ret == -ENOMSG) {
-		env_set_default("bad CRC", 0);
+		env_set_default(NULL, 0);
 		return -ENOMSG;
 	}
 
@@ -253,7 +253,7 @@
 		env_set_default(NULL, 0);
 #else
 		bootstage_error(BOOTSTAGE_ID_NET_CHECKSUM);
-		env_set_default("bad CRC", 0);
+		env_set_default(NULL, 0);
 #endif
 	} else {
 		env_load();
