--- a/arch/arm/mach-stm32mp/cpu.c	2022-07-19 09:44:18.015013000 +0200
+++ b/arch/arm/mach-stm32mp/cpu.c	2022-07-19 10:00:35.622241873 +0200
@@ -314,6 +314,7 @@
 {
 	int ret;
 	int i;
+	unsigned int seed;
 	u32 otp[3];
 	uchar enetaddr[6];
 	struct udevice *dev;
@@ -346,9 +347,15 @@
 			enetaddr[i] = ((uint8_t *)&otp)[i + 6 * index];
 
 		if (!is_valid_ethaddr(enetaddr)) {
-			log_err("invalid MAC address %d in OTP %pM\n",
-				index, enetaddr);
-			return -EINVAL;
+			pr_debug("Generating randomized MAC address\n");
+	
+			seed = get_ticks();
+
+			for (i = 0; i < 6; i++)
+				enetaddr[i] = rand_r(&seed);
+
+			enetaddr[0] &= 0xfe;	/* clear multicast bit */
+			enetaddr[0] |= 0x02;	/* set local assignment bit (IEEE802) */
 		}
 		log_debug("OTP MAC address %d = %pM\n", index, enetaddr);
 		ret = eth_env_set_enetaddr_by_index("eth", index, enetaddr);
