--- a/arch/arm/mach-stm32mp/cpu.c	2022-07-19 09:44:18.015013000 +0200
+++ b/arch/arm/mach-stm32mp/cpu.c	2022-07-19 10:00:35.622241873 +0200
@@ -598,6 +598,7 @@
 #if defined(CONFIG_NET)
 	int ret;
 	int i;
+	unsigned int seed;
 	u32 otp[2];
 	uchar enetaddr[6];
 	struct udevice *dev;
@@ -621,8 +622,15 @@
 		enetaddr[i] = ((uint8_t *)&otp)[i];
 
 	if (!is_valid_ethaddr(enetaddr)) {
-		pr_err("invalid MAC address in OTP %pM\n", enetaddr);
-		return -EINVAL;
+		pr_debug("Generating randomized MAC address\n");
+	
+		seed = get_ticks();
+
+		for (i = 0; i < 6; i++)
+			enetaddr[i] = rand_r(&seed);
+
+		enetaddr[0] &= 0xfe;	/* clear multicast bit */
+		enetaddr[0] |= 0x02;	/* set local assignment bit (IEEE802) */
 	}
 	pr_debug("OTP MAC address = %pM\n", enetaddr);
 	ret = eth_env_set_enetaddr("ethaddr", enetaddr);
