--- a/common/dfu.c	2022-08-09 12:42:36.553418000 +0200
+++ b/common/dfu.c	2022-08-09 13:04:09.143672334 +0200
@@ -20,10 +20,14 @@
 #include <usb.h>
 #include <net.h>
 
+extern int g_flashing_done_hint;
+
 int run_usb_dnl_gadget(int usbctrl_index, char *usb_dnl_gadget)
 {
 	bool dfu_reset = false;
 	int ret, i = 0;
+	int iled = 1;
+	int ibreath;
 
 	ret = usb_gadget_initialize(usbctrl_index);
 	if (ret) {
@@ -41,7 +45,22 @@
 	unsigned long start_time = get_timer(0);
 #endif
 
+	unsigned long breath_time = get_timer(0);
+	unsigned int breath_span = 68 * 1000; // Verification time required for flashing the E2.
+
+    	ibreath = gpio_request(29 /*PB13 Breathing LED*/, "BREATHING_LED");
+    	
 	while (1) {
+		if (ibreath >= 0) {
+			if (g_flashing_done_hint) {
+				if (get_timer(breath_time) > breath_span) {
+					breath_span += 500;
+					gpio_direction_output(29 /*PB13 Breathing LED*/, iled);
+					iled = iled ? 0 : 1;
+				}
+			}
+		}
+
 		if (g_dnl_detach()) {
 			/*
 			 * Check if USB bus reset is performed after detach,
