--- a/boot/pxe_utils.c	2022-07-20 10:43:25.902257000 +0200
+++ b/boot/pxe_utils.c	2022-07-20 11:11:55.509243408 +0200
@@ -20,6 +20,7 @@
 #include <linux/ctype.h>
 #include <errno.h>
 #include <linux/list.h>
+#include <linux/delay.h>
 
 #ifdef CONFIG_DM_RNG
 #include <rng.h>
@@ -27,6 +28,7 @@
 
 #include <splash.h>
 #include <asm/io.h>
+#include <asm/gpio.h>
 
 #include "menu.h"
 #include "cli.h"
@@ -1531,6 +1533,10 @@
 	struct menu *m;
 	int err;
 
+	//gpio_direction_output(83 /*PF3 Back light*/, 0);
+	
+	//udelay(50000);
+
 	if (IS_ENABLED(CONFIG_CMD_BMP)) {
 		/* display BMP if available */
 		if (cfg->bmp) {
@@ -1551,11 +1557,25 @@
 		}
 	}
 
+	udelay(400000);
+	
+	gpio_direction_output(83 /*PF3 Back light*/, 1);
+
+	udelay(1250000);
+
+	gpio_direction_output(83 /*PF3 Back light*/, 0);
+	
 	m = pxe_menu_to_menu(cfg);
 	if (!m)
 		return;
 
-	err = menu_get_choice(m, &choice);
+	menu_default_set(m, "1");
+	err = menu_default_choice(m, &choice);
+	
+	if (err != 1) {
+		err = menu_get_choice(m, &choice);
+	}
+	
 	menu_destroy(m);
 
 	/*
