--- a/cmd/pxe_utils.c	2022-07-20 10:43:25.902257000 +0200
+++ b/cmd/pxe_utils.c	2022-07-20 11:11:55.509243408 +0200
@@ -19,9 +19,11 @@
 #include <linux/ctype.h>
 #include <errno.h>
 #include <linux/list.h>
+#include <linux/delay.h>
 
 #include <splash.h>
 #include <asm/io.h>
+#include <asm/gpio.h>
 
 #include "menu.h"
 #include "cli.h"
@@ -1415,6 +1417,10 @@
 	struct menu *m;
 	int err;
 
+	gpio_direction_output(83 /*PF3 Back light*/, 0);
+	
+	udelay(50000);
+
 #ifdef CONFIG_CMD_BMP
 	/* display BMP if available */
 	if (cfg->bmp) {
@@ -1430,14 +1436,26 @@
 	}
 #endif
 
+	udelay(400000);
+	
+	gpio_direction_output(83 /*PF3 Back light*/, 1);
+
+	udelay(1250000);
+
+	gpio_direction_output(83 /*PF3 Back light*/, 0);
+
 	m = pxe_menu_to_menu(cfg);
 	if (!m)
 		return;
 
-	err = menu_get_choice(m, &choice);
+	err = 1;
+	menu_default_choice(m, &choice);
 
 	menu_destroy(m);
 
+	if (CONFIG_IS_ENABLED(CMD_CLS))
+		run_command("cls", 0);
+
 	/*
 	 * err == 1 means we got a choice back from menu_get_choice.
 	 *
