--- a/include/configs/stm32mp15_st_common.h
+++ b/include/configs/stm32mp15_st_common.h
@@ -29,23 +29,43 @@
  * for nand or spi-nand boot, distro boot with ubifs on UBI partition or
  * sdcard
  * for nor boot, distro boot on SD card = mmc0 ONLY !
+ *
+ * THOR specific PKH OTP fuses for secure ROMBoot. These are basically the SHA256 of the public key.
  */
+
+#define THOR_STM32MP1_PKH_FUSES_ENABLE 1
+#define THOR_STM32MP1_PKH_FUSES_BYTES 0x8b, 0xbd, 0xf6, 0x2b, 0x0f, 0xb6, 0x19, 0xef, 0x6c, 0x00, 0x99, 0x31, 0x07, 0xbf, 0xed, 0x4e, 0xe2, 0xf4, 0x6c, 0x46, 0x1e, 0xd6, 0x3b, 0x5e, 0xc4, 0xd8, 0xd0, 0x25, 0x05, 0xff, 0xde, 0x0e
+#define THOR_STM32MP1_PKH_FUSES_PROBE_STRING "0x8bbdf62b"    // First OTP word (First 4 fuse bytes but big endian)
+#define THOR_STM32MP1_PKH_FUSES_ADDR_STRING  "0xc4000000"    // STMP151x specific - see https://wiki.st.com/stm32mpu/wiki/How_to_use_U-Boot_stm32key_command
+
+#if THOR_STM32MP1_PKH_FUSES_ENABLE
+
+#define ST_STM32MP1_BOOTCMD "bootcmd_stm32mp=" \
+	"echo \"Checking PKH\";" \
+	"if fuse cmp 0 24 " THOR_STM32MP1_PKH_FUSES_PROBE_STRING "; then echo \"PKH ok\"; else stm32key select PKH; stm32key fuse -y " THOR_STM32MP1_PKH_FUSES_ADDR_STRING "; fi;" \
+	"echo \"Booting ${boot_device}${boot_instance}!\";" \
+	"if test ${boot_device} = usb;" \
+	"then stm32prog ${boot_device} ${boot_instance}; " \
+	"else " \
+		"run env_check;" \
+		"env set boot_targets ubifs0;" \
+		"run distro_bootcmd;" \
+	"fi;\0"
+
+#else
+
 #define ST_STM32MP1_BOOTCMD "bootcmd_stm32mp=" \
 	"echo \"Boot over ${boot_device}${boot_instance}!\";" \
-	"if test ${boot_device} = serial || test ${boot_device} = usb;" \
+	"if test ${boot_device} = usb;" \
 	"then stm32prog ${boot_device} ${boot_instance}; " \
 	"else " \
 		"run env_check;" \
-		"if test ${boot_device} = mmc;" \
-		"then env set boot_targets \"mmc${boot_instance}\"; fi;" \
-		"if test ${boot_device} = nand ||" \
-		  " test ${boot_device} = spi-nand ;" \
-		"then env set boot_targets ubifs0 mmc0; fi;" \
-		"if test ${boot_device} = nor;" \
-		"then env set boot_targets mmc0; fi;" \
+		"env set boot_targets ubifs0;" \
 		"run distro_bootcmd;" \
 	"fi;\0"
 
+#endif
+
 #define ST_STM32MP1_ALTBOOTCMD "altbootcmd=run bootcmd_stm32mp\0"
 
 #undef CFG_EXTRA_ENV_SETTINGS
