--- a/arch/arm/dts/stm32mp157a-ev1-u-boot.dtsi	2022-07-28 17:56:35.994407000 +0200
+++ b/arch/arm/dts/stm32mp157a-ev1-u-boot.dtsi	2022-07-28 18:14:24.833388640 +0200
@@ -7,19 +7,13 @@
 
 / {
 	aliases {
-		gpio26 = &stmfx_pinctrl;
 		i2c1 = &i2c2;
 		i2c4 = &i2c5;
-		pinctrl2 = &stmfx_pinctrl;
-		spi0 = &qspi;
 		usb0 = &usbotg_hs;
 	};
 };
 
 #ifndef CONFIG_TFABOOT
-&flash0 {
-	u-boot,dm-spl;
-};
 
 &qspi {
 	u-boot,dm-spl;
