--- a/plat/st/stm32mp1/bl2_plat_setup.c	2022-07-14 11:01:46.211692000 +0200
+++ b/plat/st/stm32mp1/bl2_plat_setup.c	2022-07-14 11:42:13.230237079 +0200
@@ -30,6 +30,7 @@
 #include <drivers/st/stm32mp1_clk.h>
 #include <drivers/st/stm32mp1_pwr.h>
 #include <drivers/st/stm32mp1_ram.h>
+#include <drivers/st/stm32_gpio.h>
 #include <lib/fconf/fconf.h>
 #include <lib/fconf/fconf_dyn_cfg_getter.h>
 #include <lib/mmio.h>
@@ -410,6 +411,13 @@
 	} else {
 		stm32mp_io_setup();
 	}
+	
+	/* Turn the breathing LED PB13 on. */
+ 
+ 	NOTICE("Turn the 'Breathing LED B13' on.\n");
+ 
+ 	set_gpio_config(1, 13, GPIOF_OUT_INIT_LOW, 0);
+	clk_enable(stm32_get_gpio_bank_clock(1));
 }
 
 #if STM32MP13
