--- a/common/bl_common.c	2026-04-20 15:35:56.286681053 +0200
+++ b/common/bl_common.c	2026-04-20 15:50:32.744425946 +0200
@@ -85,8 +85,6 @@
 	/* Obtain a reference to the image by querying the platform layer */
 	io_result = plat_get_image_source(image_id, &dev_handle, &image_spec);
 	if (io_result != 0) {
-		WARN("Failed to obtain reference to image id=%u (%i)\n",
-			image_id, io_result);
 		return io_result;
 	}
 
